### PR TITLE
[msal-core] Fix authority match confusion for cached token lookup

### LINKED ISSUE #2370: Still Login Redirect Loop after upgrade to 1.4.1, if previously cached tokens exists in browser store
## Library
- [x ] `msal@1.4.1`
- [ ] `@azure/msal-browser@2.x.x`
- [ ] `@azure/msal-angular@0.x.x`
- [ ] `@azure/msal-angular@1.x.x`
- [ ] `@azure/msal-angularjs@1.x.x`

## Framework
React

## Description

This is the follow up to reported and closed [MSAL 1.4.0 was released, and is causing infinite redirection loop #2185](https://github.com/AzureAD/microsoft-authentication-library-for-js/issues/2185).

The redirection loop occurs, if the browser contains already cached values from previous MSAL versions. Manual clearing of applicaition cache helps to resolve the issue. But we cannot enforce all our users do this.

In my opinion, MSAL should be able to handle this case and not blow up on migration of the library.

## Error Message

**Cache error for scope openid,profile: The cache contains multiple tokens satisfying the requirements.**

hereby MSAL log (repeats as infinity loop):

```
M32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose GetCachedToken has been called
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose No tokens found
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose Getting all cached access tokens
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose No authority passed, filtering tokens by scope
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Info-pii multiple_matching_tokens:Cache error for scope openid,profile: The cache contains multiple tokens satisfying the requirements. Call AcquireToken again providing more requirements like authority..
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose AcquireTokenSilent has been called
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose Telemetry Event started: 4339d004-04e6-4b5a-ae89-7fd1d0f4835b_0e9d0ff7-f4ad-438d-a48f-0969873b6a2e-msal.api_event
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose-pii Serialized scopes: openid profile
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose Account set from MSAL Cache
VM32 main.chunk.js:61058 MSAL: - Fri, 02 Oct 2020 07:24:54 GMT:1234-1.4.0-Verbose Response type: id_token

```

## Regression
- [x] Did this behavior work before? 
Version: 1.3.6

## MSAL Configuration

see linked #2185 

## Reproduction steps

see linked #2185 

## Expected behavior

MSAL should be able to handle multiple cache entries on migration and prevent infinite redirection loop .

## Browsers/Environment
- [x] Chrome
- [x] Firefox
- [x] Edge
- [ ] Safari
- [ ] IE
- [ ] Other (Please add browser name here)

### PR DESCRIPTION
**Pull request by hectormmg**
This PR:

-  Refactors `getCachedToken` into `getCachedAccessToken` and `getCachedIdToken` to simplify the distinction between both cases, extracting shared logic into other methods that they can each call.
- Changes `AuthCache.getIdToken` to `AuthCache.getAllIdTokens` to replicate the approach used for access token lookup for ID tokens
- Adds `AuthCacheUtils` class that specializes in filtering token cache items
- Adds extra filtering by any available authority in order to cover the case where multiple access or ID tokens for the same account are present in the cache
- Fixes an issue where ID tokens renewed in id_token token requests were not being cached as their own access token item by separating the logic in `saveAccessToken` into `saveAccessToken` and `saveIdToken` and calling each in the right cases
- Fixes #2370

### CODE REVIEW

File: lib/msal-core/src/UserAgentApplication.ts [hunk 1]
```diff
@@ -46,6 +46,7 @@ import { Constants,
 } from "./utils/Constants";
 import { CryptoUtils } from "./utils/CryptoUtils";
 import { TrustedAuthority } from "./authority/TrustedAuthority";
+import { server } from 'sinon';
```
Comment (jasonnutter) [inline lib/msal-core/src/UserAgentApplication.ts:49]: I dont think you meant to include this.

Comment (hectormmg) [reply lib/msal-core/src/UserAgentApplication.ts:49]: Yes I just noticed, fixing for AuthCache as well.

File: lib/msal-core/src/cache/AuthCache.ts [hunk 2]
```diff
@@ -3,14 +3,15 @@
...
-import { Constants, PersistentCacheKeys, TemporaryCacheKeys, ErrorCacheKeys} from "../utils/Constants";
+import { Constants, PersistentCacheKeys, TemporaryCacheKeys, ErrorCacheKeys, ServerHashParamKeys} from "../utils/Constants";
 import { AccessTokenCacheItem } from "./AccessTokenCacheItem";
 import { CacheLocation } from "../Configuration";
 import { BrowserStorage } from "./BrowserStorage";
 import { ClientAuthError } from "../error/ClientAuthError";
 import { RequestUtils } from "../utils/RequestUtils";
 import { AccessTokenValue } from "./AccessTokenValue";
 import { AccessTokenKey } from "./AccessTokenKey";
+import { Server } from 'http';
```
Comment (jasonnutter) [inline lib/msal-core/src/cache/AuthCache.ts:14]: same here

Comment (hectormmg) [reply lib/msal-core/src/cache/AuthCache.ts]: Fixed, thanks!

File: change/msal-2020-10-07-14-20-34-authority-match.json [hunk 3]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Fix issues with ID token caching and cache lookup (#2376)",
+  "packageName": "msal",
+  "email": "hectormgdev@gmail.com",
```
Comment (hectormmg) [inline change/msal-2020-10-07-14-20-34-authority-match.json:5]:
```suggestion
  "email": "hemoral@microsoft.com",
```

File: lib/msal-core/src/cache/AuthCache.ts [hunk 4]
```diff
@@ -80,6 +78,30 @@ export class AuthCache extends BrowserStorage {// Singleton
...
+    private matchKeyForType(key:string, clientId: string, homeAccountIdentifier: string, tokenType: string): Boolean {
+        const accountMatch = key.match(clientId) && key.match(homeAccountIdentifier);
+        
+        switch (tokenType) {
+            case ServerHashParamKeys.ACCESS_TOKEN:
+                // Cache item is an access token if scopes are included in the cache item key
+                return Boolean(accountMatch && key.match(Constants.scopes));
+            case ServerHashParamKeys.ID_TOKEN:
+                // Cache item is an ID token if scopes are NOT included in the cache item key
+                return Boolean(accountMatch && !key.match(Constants.scopes));
```
Comment (jasonnutter) [inline lib/msal-core/src/cache/AuthCache.ts:99]: Nitpick: `return !!(accountMatch && !key.match(Constants.scopes));`

**Review by jasonnutter — APPROVED** [review]
One thing otherwise looks good!

File: lib/msal-core/src/cache/AuthCache.ts (hunk 4 above)
Comment (hectormmg) [reply lib/msal-core/src/cache/AuthCache.ts]: Done, thanks!

File: change/msal-2020-10-07-14-20-34-authority-match.json [hunk 5]
```diff
@@ -0,0 +1,8 @@
+{
+  "type": "patch",
+  "comment": "Fix issues with ID token caching and cache lookup (#2376)",
+  "packageName": "msal",
+  "email": "hemoral@microsoft.com",
+  "dependentChangeType": "patch",
```
Comment (jasonnutter) [inline change/msal-2020-10-07-14-20-34-authority-match.json:6]:
```suggestion
  "dependentChangeType": "none",
```

File: lib/msal-core/src/UserAgentApplication.ts [hunk 6]
```diff
@@ -1401,118 +1440,37 @@ export class UserAgentApplication {
...
+     * 
+     * @param serverAuthenticationRequest 
+     * @param account 
+     * @param scopes 
+     */
     private getCachedAccessToken(serverAuthenticationRequest: ServerRequestParameters, account: Account, scopes: string[]): AuthResponse {
-        // filter by clientId and account
-        const tokenCacheItems = this.cacheStorage.getAllAccessTokens(this.clientId, account ? account.homeAccountIdentifier : null);
+        this.logger.verbose("Getting all cached tokens of type Access Token");
+        const tokenCacheItems = this.cacheStorage.getAllTokensByType(this.clientId, account ? account.homeAccountIdentifier : null, ServerHashParamKeys.ACCESS_TOKEN);
```
Comment (tnorling) [inline lib/msal-core/src/UserAgentApplication.ts:1455]: Reason for this change? `getAllAccessTokens` seemed more descriptive

**Review by tnorling — APPROVED** [review]
One comment, otherwise looks good!

File: lib/msal-core/src/UserAgentApplication.ts (hunk 6 above)
Comment (hectormmg) [reply lib/msal-core/src/UserAgentApplication.ts]: Changed it hoping I'd be able to factor the beginning of these methods into a single shared method for IdTokens and AccessTokens, but it turned out to not be trivial. Switched back to using the type-specific methods.

File: lib/msal-core/test/UserAgentApplication.spec.ts [hunk 7]
```diff
@@ -40,7 +40,7 @@ type kv = {
     [key: string]: string;
 };
 
-describe("UserAgentApplication.ts Class", function () {
+describe.only("UserAgentApplication.ts Class", function () {
```
Comment (jasonnutter) [inline lib/msal-core/test/UserAgentApplication.spec.ts:43]: `.only`

Comment (jasonnutter) [inline lib/msal-core/test/UserAgentApplication.spec.ts:43]:
```suggestion
describe("UserAgentApplication.ts Class", function () {
```

Comment (hectormmg) [reply lib/msal-core/test/UserAgentApplication.spec.ts]: Thanks for the catch!

**Pull request merged**